### PR TITLE
[xbmc] Link against curl instead of using dlopen.

### PR DESCRIPTION
**Pull request by Paxxi**
This replaces our dlopen usage of curl with regular linking

Runtime tested on windows x64
## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by MilhouseVH** [comment]
Getting these build errors with this PR and LibreELEC: http://ix.io/wDS

**Comment by MilhouseVH** [comment]
This patch fixes the Linux build issue: http://ix.io/wEG

I've now built LibreELEC for x86_64 and this seems OK.

**Comment by Rechi** [comment]
`#include <string>` should be added in `DllLibCurl.h` instead of including it before `DllLibCurl.h`

**Comment by MilhouseVH** [comment]
@Rechi I guessed which file to add it to. :)

But adding it in either file should fix the build issue. You fellas know best what to do.

**Comment by Paxxi** [comment]
thanks for testing @MilhouseVH . I didn't build test it on linux before pushing so not unexpected 😄

File: xbmc/filesystem/DllLibCurl.cpp [hunk 1]
```diff
@@ -59,38 +60,124 @@ unsigned long ssl_thread_id(void)
...
-    return true;
-  }
+  curl_global_cleanup();
+}
 
-  /* we handle this ourself */
-  DllDynamic::EnableDelayedUnload(false);
-  if (!DllDynamic::Load())
-    return false;
+CURL_HANDLE * DllLibCurl::easy_init()
```
Comment (notspiff) [inline xbmc/filesystem/DllLibCurl.cpp]: do we really need this wrapper class now ?

Comment (Paxxi) [reply xbmc/filesystem/DllLibCurl.cpp]: At this stage not really. I was thinking of making the class into a more designed wrapper that keeps track of easy/multi handles and encapsulates everything curl in one place.
I also want to get rid of the state in curl global but it requires more work so it'll have to wait for a separate pr.

Comment (notspiff) [reply xbmc/filesystem/DllLibCurl.cpp]: okay, if it's a todo..

**Review by notspiff — APPROVED** [review]

File: xbmc/filesystem/DllLibCurl.cpp [hunk 2]
```diff
@@ -37,170 +38,228 @@ extern "C"
...
+    if (!m_sslLockArray)
+      return;
+
+    if (mode & CRYPTO_LOCK)
+      m_sslLockArray[type]->lock();
+    else
+      m_sslLockArray[type]->unlock();
+  }
+
+  unsigned long ssl_thread_id(void) { return (unsigned long)CThread::GetCurrentThreadId(); }
```
Comment (Rechi) [inline xbmc/filesystem/DllLibCurl.cpp]: c++ style cast

File: xbmc/filesystem/DllLibCurl.cpp [hunk 3]
```diff
@@ -37,170 +38,228 @@ extern "C"
...
   VEC_CURLSESSIONS::iterator it = m_sessions.begin();
-  while(it != m_sessions.end())
+  while (it != m_sessions.end())
   {
-    if( !it->m_busy && (XbmcThreads::SystemClockMillis() - it->m_idletimestamp) > idletime )
+    if (!it->m_busy && (XbmcThreads::SystemClockMillis() - it->m_idletimestamp) > idletime)
     {
-      CLog::Log(LOGINFO, "%s - Closing session to %s://%s (easy=%p, multi=%p)\n", __FUNCTION__, it->m_protocol.c_str(), it->m_hostname.c_str(), (void*)it->m_easy, (void*)it->m_multi);
+      CLog::Log(LOGINFO, "%s - Closing session to %s://%s (easy=%p, multi=%p)\n", __FUNCTION__,
+                it->m_protocol.c_str(), it->m_hostname.c_str(), (void*)it->m_easy,
```
Comment (Rechi) [inline xbmc/filesystem/DllLibCurl.cpp]: c++ style cast

File: xbmc/filesystem/DllLibCurl.cpp [hunk 4]
```diff
@@ -37,170 +38,228 @@ extern "C"
...
+                (void*)it->m_multi);
```
Comment (Rechi) [inline xbmc/filesystem/DllLibCurl.cpp]: c++ style cast

**Review by Rechi — CHANGES_REQUESTED** [review]

**Comment by Paxxi** [comment]
So after discussion on Slack I changed curl to be built statically, not sure if it's done the right way.

File: tools/depends/target/cmakebuildsys/Makefile [hunk 5]
```diff
@@ -18,7 +18,7 @@ BUILD_DIR=$(CMAKE_SOURCE_DIR)/build
 
 all:
 	mkdir -p $(BUILD_DIR)
-	cd $(BUILD_DIR); $(CMAKE) -DCMAKE_BUILD_TYPE=$(Configuration) -DENABLE_INTERNAL_CROSSGUID=OFF -DENABLE_INTERNAL_FFMPEG=OFF $(CMAKE_EXTRA_ARGUMENTS) $(CMAKE_SOURCE_DIR)
+	cd $(BUILD_DIR); $(CMAKE) -DCMAKE_BUILD_TYPE=$(Configuration) -DENABLE_INTERNAL_CROSSGUID=OFF -DENABLE_INTERNAL_FFMPEG=OFF -DHAS_CURL_STATIC=ON $(CMAKE_EXTRA_ARGUMENTS) $(CMAKE_SOURCE_DIR)
```
Comment (Rechi) [inline tools/depends/target/cmakebuildsys/Makefile]: `HAS_CURL_STATIC` is autodetected or did you face issues?

Comment (Paxxi) [reply tools/depends/target/cmakebuildsys/Makefile]: it failed without it, saying it couldn't find curl so. after it found curl

Comment (Rechi) [reply tools/depends/target/cmakebuildsys/Makefile]: which platform?

Comment (Paxxi) [reply tools/depends/target/cmakebuildsys/Makefile]: android/arm

**Comment by Rechi** [comment]
wrong revision was build
jenkins build this please

**Comment by Rechi** [comment]
wrong revision was build
jenkins build this please

**Pull request merged**